src/mongo/db/encryption/key_id.h

```diff
@@ -0,0 +1,301 @@
+    code of portions of this program with the OpenSSL library under certain
+    conditions as described in each individual source file and distribute
+    linked combinations including the program with the OpenSSL library. You
+    must comply with the Server Side Public License in all respects for
+    all of the code used other than as permitted herein. If you modify file(s)
+    with this exception, you may extend this exception to your version of the
+    file(s), but you are not obligated to do so. If you do not wish to do so,
+    delete this exception statement from your version. If you delete this
+    exception statement from all source files in the program, then also delete
+    it in the license file.
+======= */
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+namespace mongo {
+class BSONObj;
+class BSONObjBuilder;
+class EncryptionGlobalParams;
+
+namespace encryption {
+class KeyIdConstVisitor;
+class ReadKey;
+
+class KeyId {
+public:
+    // @todo Consider introducing the macro which, given a class name, would
+    // declare all the six special member functions with default definitions,
+    // making the destructor virtual.
+    // Using the macro may eliminate boilerplate code.
+    // @see the `KeyIdConstVisitor` classes where
+    // boilerplate code is about half of the definition;
+    virtual ~KeyId() = default;
+    KeyId() = default;
+    KeyId(const KeyId&) = default;
+    KeyId(KeyId&&) = default;
+    KeyId& operator=(const KeyId&) = default;
+    KeyId& operator=(KeyId&&) = default;
+
+    virtual std::unique_ptr<KeyId> clone() const = 0;
+
+    /// @brief Reads an encryption key identifier (if any) from the storage
+    /// engine encryption options.
+    ///
+    /// @param options storage engine encryption options
+    ///
+    /// @returns pointer to the read key identifier, if any
+    ///
+    /// @throws std::runtime_error in case of invalid options format
+    static std::unique_ptr<KeyId> fromStorageEngineEncryptionOptions(const BSONObj& options);
+    /// @brief Serializes the encryption key identifier to the storage engine
+    /// encryption options, which, in turn, are kept in storage engine metadata.
+    void serializeToStorageEngineEncryptionOptions(BSONObjBuilder* b) const;
+    virtual bool needsSerializationToStorageEngineEncryptionOptions() const noexcept = 0;
+
+    /// @brief Makes the class and its subclasses loggable with LOGV2.
+    virtual void serialize(BSONObjBuilder* b) const = 0;
+
+    virtual void accept(KeyIdConstVisitor& v) const = 0;
+
+    /// @brief Returns the type of the key management facility suitable for
+    /// keeping the key this identifier points to.
+    virtual const char* facilityType() const noexcept = 0;
+
+private:
+    /// @brief Returns field name under which this key identifier is serialized
+    /// to the storage engine encryption opitons.
+    ///
+    /// @note sseo stands for storage engine encryption options.
```

Comment:
Well, if I understand correctly, we have secrets engine configuration here (like path, version), which is used to read correct encryption key from the secrets engine. If that is correct then this is not "storage engine encryption options".

Reply:
or maybe even "key management facility config" ?

Reply:
> avoid ambiguity with "real" storage engines (like wiredTiger)

By using "storage engine encryption options", I meant the "real" storage engine, WiredTiger in particular. When `mongod` prepares to read the data at rest, it first reads what is called storage engine metadata from the `storage.bson` file. A part of that metadata is storage engine options, hence the member function [`StorageEngineMetadata::getStorageEngineOptions`](https://github.com/percona/percona-server-mongodb/blob/master/src/mongo/db/storage/storage_engine_metadata.h#L81) which returns a reference to a BSON object. It is its subobject under the "encryption" field that contains the identifier of the key the data at rest is encrypted with. This key relates to the encryption of the data the "real" storage engine manages, that's why the subobject under the "encryption" field is called "storage engine encryption options". Please note, that the subobject may, in theory, comprise other data in the future as well, thus "options" is used, not just "key identifier".

On the other hand, I don't think "key management facility config" is a precise term for a BSON subobject consisting exclusively of encryption key identifier, because one can't really "configure" anything with it. No one can't even guarantee that the key with a particular identifier exists on a remote "key management facility", be it Vault or KMIP server.

Any other opinions?

Reply:
> typo: sseo/seeo

fixed